### ISSUE TITLE
Parser fails on large files due to arithmetic overflow.

### ISSUE THREAD

**Issue by alh-solidatus** [justinwilaby/sax-wasm#55]
I have been trying to use the library to process large XML files (~100 GB) which do not fit in memory of the machine. I copied the example code for node-based streaming files in the README.

It seems that the WASM code does not wait for written data to be processed when writing data and instead allows the buffer to grow infinitely, eating the system memory. Is there any way to make this block here?

Moreover, the processes crashes before memory exhaustion is reached with this error: 

```
node_modules/sax-wasm/lib/saxWasm.js:186
            const uint8array = new Uint8Array(this.wasmSaxParser.memory.buffer, ptr, len);
                               ^

RangeError: Start offset -2147483568 is outside the bounds of the buffer
    at new Uint8Array (<anonymous>)
```

I'm not exactly sure of what's going on here but it looks like the buffer may have grown too big and the index within it has exceeded s32_max, causing a failure. Is this a bug or am I doing things wrong - would be good to have documentation if so.

**Comment by justinwilaby** [comment]
Interesting. 

Each write starts at index zero within the wasm mem buffer and is processed synchronously by the parser from that point on. This guarantees that each chunk from the filesystem is not aggregated in memory. 

The parser does use a stack to store information about tags that are encountered. Open tags push to the stack, while close tags pop from the stack. These do take up internal memory and may be the culprit here if the XML document contains very large text blocks combined with several hundred attributes with a very high nest level. 

The negative offset value is suspicious since this should be unsigned in every case. Javascript uses Two's Compliment to represent negative numbers and it seems this is at play. For example, the binary two's complement representation of  -2147483568 is `0b10000000000000000000000001010000` with the first significant bit representing the sign. The uint representation is `2147483728` which is probably the correct pointer and should not produce this error.

I'll look into this a bit further. Any idea on how to obtain a ready-made 100gb XML?

**Comment by alh-solidatus** [comment]
Thanks for the reply! I'm not sure where to get a suitably giant test file - unfortunately I can't share any of the ones I've been working with; but there shouldn't be anything special about them - the typical nesting level is about 10 levels, max 20 and some larger blocks of text, but on the order of 10 kB not really anything massive. I'll have a look for/generate some test data and see if I can reproduce with something publicly available.

**Comment by alh-solidatus** [comment]
I've reproduced it on some publicly available data - a 6.5 GB dump from wikipedia: https://ftp.acc.umu.se/mirror/wikimedia.org/dumps/enwiki/20220101/enwiki-20220101-abstract.xml.gz

This is the core of the code I was using to reproduce: https://gist.github.com/alh-solidatus/0bf1b75af70eadbc40df81f27cad8070

**Comment by justinwilaby** [comment]
@alh-solidatus - 

#56  should fix the issue and give  you a 30+% increase in performance. Can you verify this fix?

**Comment by alh-solidatus** [comment]
I've tested it and it works perfectly, thanks!

**Comment by justinwilaby** [comment]
Fixed by #56
